src/vcpkg/vcpkgpaths.cpp

```diff
@@ -1008,20 +1009,40 @@ namespace vcpkg
         }
     }
 
-    ExpectedL<bool> VcpkgPaths::git_port_has_local_changes(StringView port_name) const
+    ExpectedL<std::vector<std::string>> VcpkgPaths::git_builtin_ports_status() const
     {
         const auto cmd = git_cmd_builder({}, {})
                              .string_arg("status")
                              .string_arg("--porcelain=v1")
                              .string_arg("--")
-                             .string_arg(Strings::concat("ports/", port_name));
+                             .string_arg("ports/");
+
         auto output = cmd_execute_and_capture_output(cmd);
-        if (output.exit_code == 0)
+        if (output.exit_code != 0)
         {
-            return !output.output.empty();
+            return msg::format(msgGitCommandFailed, msg::command_line = cmd.command_line())
+                .appendnl()
+                .append_raw(output.output);
         }
 
-        return msg::format(msgGitCommandFailed, msg::command_line = cmd.command_line());
+        // Output of git status --porcelain=v1 is in the form: XY PATH
+        // where X represents the index status and Y the working tree status
+        // for now we ignore the status and only care about the paths
+        // static constexpr size_t STATUS = 0;
+        static constexpr size_t PATH = 1;
+
+        std::vector<std::string> status_paths;
+        auto lines = Strings::split(output.output, '\n');
+        for (auto&& line : lines)
+        {
+            auto&& values = Strings::split(line, ' ');
+            if (values.size() != 2)
+            {
+                return msg::format(msgGitUnexpectedCommandOutput).appendnl().append_raw(output.output);
+            }
+            status_paths.emplace_back(values[PATH]);
+        }
```

Comment:
I think this would be more clearly correct if it actually did the parsing rather than tried to reconstitute it from splits etc. As in:

```
struct GitStatusLine {
    char index_status;
    char working_tree_status; // or enum classes for these
    std::string path;

    StringView port_name() const {
        static constexpr StringLiteral prefix = "/ports/";
        static constexpr min_path_size = sizeof("/ports/x/") - 1;
        if (path.size() > min_path_size && Strings::starts_with(prefix, path)) {
            auto without_port_prefix = StringView{path}.substr(prefix.size());
            auto slash = without_port_prefix.find('/');
            return without_port_prefix.substr(0, slash);
        }

        return StringView{};
    }
};

std::vector<GitStatusLine> results;
ParserBase parser(output.output, "git status output");
for (;;) {
    parser.skip_whitespace();
    if (parser.at_eof()) { return results; }
    GitStatusLine result;
    result.index_status = parser.cur(); // if you picked enum classes for these, try ParserBase::try_match_keyword ?
    result.working_tree_status = parser.next();
    parser.require_character(' ');
    result.path = parser.match_until(ParserBase::is_lineend).to_string();
    results.emplace_back(std::move(result));
}
```

or similar?

Reply:
I'll give it a go, thanks Billy!